lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -5819,44 +6061,31 @@ void Differentiation::run() {
     return;
   }
 
-  // A global differentiation context.
-  ADContext context(*this);
-
-  // For every `[differentiable]` attribute, create a differentiation task. If
-  // the attribute has a primal and adjoint, this task will not synthesize
-  // anything, but it's still needed as a lookup target.
-  for (auto &fnAndAttr : diffAttrs) {
-    context.registerDifferentiationTask(
-        fnAndAttr.first, fnAndAttr.second->getIndices(),
-        DifferentiationInvoker(fnAndAttr.second, fnAndAttr.first));
-  }
-
-  bool errorProcessingAutoDiffInsts = false;
-  for (auto *adfi : autodiffInsts)
-    errorProcessingAutoDiffInsts |= context.processAutoDiffFunctionInst(adfi);
-
-  auto cleanUp = [&]() {
-    for (auto &task : context.getDifferentiationTasks())
-      context.clearTask(task.get());
-  };
-
-  // Run primal generation for newly created differentiation tasks. If any error
-  // occurs, continue on the tasks that are still good.
-  PrimalGen primalGen(context);
-  bool primalGenError = primalGen.run();
-
-  // Run adjoint generation for differentiation tasks. If any error occurs, back
-  // out.
-  AdjointGen adjointGen(context);
-  if (adjointGen.run() || primalGenError) {
-    cleanUp();
-    return;
-  }
+  bool errorOccurred = false;
 
-  // If there was any error that occurred during `autodiff_function` instruction
-  // processing, back out.
-  if (errorProcessingAutoDiffInsts) {
-    cleanUp();
+  // Process all `[differentiable]` attributes.
+  for (auto invokerInfo : context.getInvokers()) {
+    auto *attr = invokerInfo.first;
+    auto *original = attr->getOriginal();
+    auto invoker = invokerInfo.second;
+    errorOccurred |=
+        context.processDifferentiableAttribute(original, attr, invoker);
+  }
+
+  // Iteratively process `autodiff_function` instruction worklist.
+  while (!context.getAutoDiffFunctionInsts().empty()) {
+    auto *adfi = context.getAutoDiffFunctionInsts().back();
+    context.getAutoDiffFunctionInsts().pop_back();
+    // Skip instructions that have been set to nullptr by
+    // `processAutoDiffFunctionInst`.
+    if (adfi)
```

Comment:
Maybe an early return here?

```suggestion
    if (!adfi) continue;
```

Reply:
I'm a fan of `if (!adfi) continue;` just for the uniform indentation. 😃